kmath-core/src/commonMain/kotlin/space/kscience/kmath/tensors/core/DoubleTensorAlgebra.kt

```diff
@@ -241,8 +241,84 @@ public open class DoubleTensorAlgebra : TensorPartialDivisionAlgebra<Double, Dou
         TODO("Not yet implemented")
     }
 
+    private fun DoubleTensor.dotTwoDimensionalTensors(other: DoubleTensor): DoubleTensor {
+        if (this.shape.size > 2 || other.shape.size > 2) {
+            throw RuntimeException("Both tensors must have a maximum of 2 dimensions")
+        }
+
+        if (this.shape[1] != other.shape[0]) {
+            throw RuntimeException("Tensors dot operation dimension mismatch: " +
+                    "(${this.shape[0]}, ${this.shape[1]}) x (${other.shape[0]}, ${other.shape[1]})")
+        }
+
+        val l = this.shape[0]
+        val m = this.shape[1]
+        val n = other.shape[1]
+
+        val res = DoubleTensor(intArrayOf(l, n), DoubleArray(l * n))
+
+        for (i in 0 until l) {
+            for (j in 0 until n) {
+                var curr = 0.0
+                for (k in 0 until m) {
+                    val ik = this.linearStructure.offset(intArrayOf(i, k))
+                    val kj = other.linearStructure.offset(intArrayOf(k, j))
+                    curr += this.buffer.array()[ik] * other.buffer.array()[kj]
+                }
+                val linearIndex = res.linearStructure.offset(intArrayOf(i, j))
+                res.buffer.array()[linearIndex] = curr
+            }
+        }
+        return res
+    }
+
     override fun DoubleTensor.dot(other: DoubleTensor): DoubleTensor {
-        TODO("Alya")
+        if (this.shape.size == 1 && other.shape.size == 1) {
+            return DoubleTensor(intArrayOf(1), doubleArrayOf(this.times(other).buffer.array().sum()))
+        }
+
+        var newThis = this.copy()
+        var newOther = other.copy()
+        if (this.shape.size == 1) {
+            newThis = this.view(intArrayOf(1) + this.shape)
+        }
+        if (other.shape.size == 1) {
+            newOther = other.view(other.shape + intArrayOf(1) )
+        }
+
+        val broadcastTensors = broadcastOuterTensors(newThis, newOther)
+        newThis = broadcastTensors[0]
+        newOther = broadcastTensors[1]
+
+        val l = newThis.shape[newThis.shape.size - 2]
+        val m1= newThis.shape[newThis.shape.size - 1]
+        val m2 = newOther.shape[newOther.shape.size - 2]
+        val n = newOther.shape[newOther.shape.size - 1]
+        if (m1 != m2) {
+            throw RuntimeException("Tensors dot operation dimension mismatch: ($l, $m1) x ($m2, $n)")
```

Comment:
For such preconditions, IllegalStateException or IllegalArgumentException should be thrown.